### ISSUE TITLE
Grouped aggregates and non-terminal reductions

### ISSUE THREAD

**Issue by cxdorn** [typedb/typedb#7156]
## Problem to Solve

> ***Grouped aggregations are a core data analysis command***

In its current specification, TypeDB "almost" support this features, but it could be made much more convenient.

## Current Workaround

For example, consider counting posts from a specific country by tag. Using #7038 we can write:

```
with fun count_posts($tag: tag, $country: country) -> long:
  match
    $post isa post; 
    location_check($post, $country) == true;
    (post: $post, tag: $tag) isa tagging;
  return count(post);
match
  $country isa country, has name "Australia";
  $post isa post; 
  location_check($post, $country) == true;
  (post: $post, tag: $tag) isa tagging;
select $tag, $country;
distinct;
  $count = count_posts($tag, $country)
fetch:
  "tag": $tag.text;
  "post count": $count;
```

But as this simple example shows, the current work around requires code duplication.

## Proposed Solution

Extend the role of the `reduce` operator in pipelines (in particular, make it non-terminal). More specifically:
* introduce `as` in `reduce` statement to specify the variables in which to store values for the next stage of the pipeline,
* introduce `group(...)` as a way to specify variables by which to group the incoming stream before reducing _each group_.

With this, the previous example becomes:

```
match
  $country isa country, has name "Australia";
  $post isa post; 
  location_check($post, $country) == true;
  (post: $post, tag: $tag) isa tagging;
reduce group($tag), count($post) as $count;
fetch:
  "tag": $tag.text;
  "post count": $count;
```

## Additional Information

References:
https://duckdb.org/2022/03/07/aggregate-hashtable.html
https://www.ibm.com/docs/en/psfa/7.1.0?topic=functions-grouped-aggregates

**Comment by krishnangovindraj** [comment]
What's the body of `location_check` like? 
I'm wary of the provided example because it has to enumerate all posts. 
Would it be more realistic to replace `location_check($post, $country)` with `$country in locations_of($post);` ?

**Comment by cxdorn** [comment]
@krishnangovindraj whether or not it'll enumerate all posts depends on the query planner. In my mind, the `locations_check` conditions would just be inlined... but whether or not that's the case is really besides the point of the issue :P

**Comment by krishnangovindraj** [comment]
Right, it is besides the point because you're grouping by tag.
I was going to suggest a syntax closer to: 
```
match
  $country isa country, has name "Australia";   
  $tag isa tag;
  $post-count = {
    $post isa post; 
    ($post, $country) isa located_in;
    (post: $post, tag: $tag) isa tagging;
    return count($post) ;
} ;
fetch:
  "tag": $tag.text;
  "post count": $count;
```

A bit like lambdas. This isn't ideal in this case, because I generate all `tag`s.

**Comment by cxdorn** [comment]
Yes it was an important detail that we are only considering tags of posts from the country ... that's the origin of the code duplication. I.e. the general problem is:
1. have a connected pattern that mutually constraints variables
2. count results grouped by one of the vars in that pattern

**Comment by flyingsilverfin** [comment]
We have settled on the syntax ```reduce $c = count($a), $d = count($b) within ($x, $y, $z);```

**Comment by krishnangovindraj** [comment]
Does this replace the reduce at the end of functions? 
This returns a stream, whereas a reduce at the end of a function returns a single value.

**Comment by cxdorn** [comment]
@krishnangovindraj there is no "explicit" `reduce` at the end of function, just a `return` which can return either a concept row or a stream of concepts rows.

`reduce` is more powerful in other regards: it has grouping as discussed above. With grouping `reduce` ***does*** return a stream!

In summary: `return` and `reduce` are not exactly the same.

**Comment by flyingsilverfin** [comment]
Implemented, forgot to close!
